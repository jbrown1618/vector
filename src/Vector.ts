import { AbstractVector } from './AbstractVector';
import { Matrix, MatrixData } from './Matrix';

export type VectorData = Array<number>;

export class Vector implements AbstractVector<Vector> {
  private readonly _data: VectorData;

  private constructor(contents: VectorData) {
    this._data = contents;
  }

  static fromData(contents: VectorData): Vector {
    return new Vector(contents);
  }

  static fromValues(...values: VectorData): Vector {
    return Vector.fromData(values);
  }

  public getData(): VectorData {
    return [...this._data];
  }

  public getEntry(index: number): number {
    if (index >= this.getDimension()) {
      throw new Error('Index out of bounds');
    }
    return this._data[index];
  }

  public add(other: Vector): Vector {
    this.checkForDimensionMismatch(other);

    const newContents = this._data.map((entry, i) => {
      return entry + other._data[i];
    });
    return new Vector(newContents);
  }

  public scalarMultiply(scalar: number): Vector {
    const newContents = this._data.map(entry => entry * scalar);
    return new Vector(newContents);
  }

  public innerProduct(other: Vector): number {
    const degenerateInnerProduct = 0;
    this.checkForDimensionMismatch(other);

    return this._data
      .map((entry, i) => entry * other._data[i])
      .reduce((sum, current) => sum + current, degenerateInnerProduct);
  }

  public outerProduct(other: Vector): Matrix {
    const matrixData: MatrixData = [];

    if (this.getDimension() === 0 || other.getDimension() === 0) {
      return Matrix.fromData([]);
    }

    this.getData().forEach((thisValue, rowIndex) => {
      matrixData[rowIndex] = [];
      other.getData().forEach((otherValue, columnIndex) => {
        matrixData[rowIndex][columnIndex] = thisValue * otherValue;
      });
    });
    return Matrix.fromData(matrixData);
  }

  public getDimension(): number {
    return this._data.length;
  }

  public approxEquals(other: Vector): boolean {
    if (this.getDimension() !== other.getDimension()) {
      return false;
    }

    return this.getData()
      .map((entry, i) => Math.abs(entry - other.getEntry(i)) < 0.0000001)
      .reduce((all, current) => all && current, true);
  }

  private checkForDimensionMismatch(other: Vector) {
    if (this.getDimension() != other.getDimension()) {
      throw Error('Dimension mismatch');
    }
  }
}
