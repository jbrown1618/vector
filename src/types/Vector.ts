import { ScalarContainer } from './ScalarContainer';
import { Matrix } from './Matrix';

export type VectorData<ScalarType> = Array<ScalarType>;

export interface Vector<ScalarType> extends ScalarContainer<ScalarType> {
  getData(): VectorData<ScalarType>;

  getEntry(index: number): ScalarType;

  add(other: Vector<ScalarType>): Vector<ScalarType>;

  scalarMultiply(scalar: ScalarType): Vector<ScalarType>;

  innerProduct(other: Vector<ScalarType>): ScalarType;

  outerProduct(other: Vector<ScalarType>): Matrix<ScalarType>;

  getDimension(): number;

  equals(other: Vector<ScalarType>): boolean;
}
