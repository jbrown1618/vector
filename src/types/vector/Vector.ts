import { Matrix } from '../matrix/Matrix';
import { MatrixBuilder } from '../matrix/MatrixBuilder';
import { ScalarOperations } from '../scalar/ScalarOperations';
import { VectorBuilder } from './VectorBuilder';

/**
 * The data stored in a {@link Vector} represented as a map
 * @public
 */
export type VectorData<S> = readonly S[];

/**
 * @internal
 */
export interface VectorConstructor<S, V extends Vector<S>> {
  new (data: VectorData<S>): V;
  ops(): ScalarOperations<S>;
  builder(): VectorBuilder<S, V>;
}

/**
 * A generalized Vector - one of the core data types
 * @public
 */
export interface Vector<S> {
  /**
   * Yields a `ScalarOperations` object which will allow consumers to work generically
   * with the scalars contained in the vector.
   * @public
   */
  ops(): ScalarOperations<S>;

  /**
   * Yields a `VectorBuilder` which will build new vectors of the same type
   * @public
   */
  builder(): VectorBuilder<S, Vector<S>>;

  /**
   * Yields a `MatrixBuilder` which will build new matrices of a compatible type
   * @public
   */
  matrixBuilder(): MatrixBuilder<S, Vector<S>, Matrix<S>>;

  /**
   * @returns The contents of the vector as an array
   * @public
   */
  toArray(): S[];

  /**
   * @returns The contents of the vector as a map of indices to nonzero values
   * @public
   */
  getSparseData(): Map<number, S>;

  /**
   * @param index - The index of the entry to retrieve
   * @returns The entry located at `index`
   * @public
   */
  getEntry(index: number): S;

  /**
   * Implements vector addition
   *
   * @param other - The vector to add
   * @returns The vector sum
   * @public
   */
  add(other: Vector<S>): Vector<S>;

  /**
   * Implements vector multiplication by a scalar
   *
   * @param scalar - The scalar by which to multiply
   * @returns The product
   * @public
   */
  scalarMultiply(scalar: S): Vector<S>;

  /**
   * @param other - The vector with which to calculate an inner product
   * @returns The inner product
   * @public
   */
  innerProduct(other: Vector<S>): S;

  /**
   * @param u - The vector on which to project this
   * @returns This vector projected onto `u`
   * @public
   */
  projectOnto(u: Vector<S>): Vector<S>;

  /**
   * @param other - The vector with which to calculate an outer product
   * @returns The outer product
   * @public
   */
  outerProduct(other: Vector<S>): Matrix<S>;

  /**
   * @returns The dimension of the vector
   * @public
   */
  getDimension(): number;

  /**
   * @param other - The vector against which to compare
   * @returns - `true` if `this` is equal to `other`
   * @public
   */
  equals(other: Vector<S>): boolean;
}
