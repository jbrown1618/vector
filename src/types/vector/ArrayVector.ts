import { assertHomogeneous, assertValidVectorIndex } from '../../utilities/ErrorAssertions';
import { Matrix } from '../matrix/Matrix';
import { MatrixBuilder } from '../matrix/MatrixBuilder';
import { ScalarOperations } from '../scalar/ScalarOperations';
import { Vector, VectorData } from './Vector';
import { VectorBuilder } from './VectorBuilder';

/**
 * Implements `Vector` with an array of values.
 *
 * @remarks
 * Subclasses must specify the usual scalar operations on their contents.
 *
 * @public
 */
export abstract class ArrayVector<S> implements Vector<S> {
  private readonly _data: VectorData<S>;

  protected constructor(data: VectorData<S>) {
    this._data = Object.freeze(data);
  }

  public abstract ops(): ScalarOperations<S>;
  public abstract builder(): VectorBuilder<S, Vector<S>>;
  public abstract matrixBuilder(): MatrixBuilder<S, Vector<S>, Matrix<S>>;

  /**
   * {@inheritDoc Vector.getEntry}
   */
  public getEntry(index: number): S {
    assertValidVectorIndex(this, index);
    return this._data[index];
  }

  /**
   * {@inheritDoc Vector.add}
   */
  public add(other: Vector<S>): Vector<S> {
    assertHomogeneous([this, other]);

    const newData = this.toArray().map((entry, index) =>
      this.ops().add(entry, other.getEntry(index))
    );

    return this.builder().fromArray(newData);
  }

  /**
   * {@inheritDoc Vector.equals}
   */
  public equals(other: Vector<S>): boolean {
    if (this.getDimension() !== other.getDimension()) {
      return false;
    }

    return this._data
      .map((entry, i) => this.ops().equals(entry, other.getEntry(i)))
      .reduce((all, current) => all && current, true);
  }

  /**
   * {@inheritDoc Vector.innerProduct}
   */
  public innerProduct(other: Vector<S>): S {
    assertHomogeneous([this, other]);

    return this._data
      .map((entry, index) =>
        this.ops().multiply(entry, this.ops().conjugate(other.getEntry(index)))
      )
      .reduce(this.ops().add, this.ops().zero());
  }

  /**
   * {@inheritDoc Vector.outerProduct}
   */
  public outerProduct(other: Vector<S>): Matrix<S> {
    const matrixData: S[][] = [];

    if (this.getDimension() === 0 || other.getDimension() === 0) {
      return this.matrixBuilder().fromArray(matrixData);
    }

    this.toArray().forEach((thisValue, rowIndex) => {
      matrixData[rowIndex] = [];
      other.toArray().forEach((otherValue, columnIndex) => {
        matrixData[rowIndex][columnIndex] = this.ops().multiply(thisValue, otherValue);
      });
    });

    return this.matrixBuilder().fromArray(matrixData);
  }

  /**
   * {@inheritDoc Vector.projectOnto}
   */
  public projectOnto(u: Vector<S>): Vector<S> {
    const oneOverUDotU = this.ops().getMultiplicativeInverse(u.innerProduct(u));
    if (oneOverUDotU === undefined) {
      throw Error('TODO - cannot project onto the zero vector');
    }

    const uDotV = u.innerProduct(this);
    const magnitudeOfProjection = this.ops().multiply(uDotV, oneOverUDotU);
    return u.scalarMultiply(magnitudeOfProjection);
  }

  /**
   * {@inheritDoc Vector.scalarMultiply}
   */
  public scalarMultiply(scalar: S): Vector<S> {
    return this.builder().fromArray(
      this.toArray().map(entry => this.ops().multiply(entry, scalar))
    );
  }

  /**
   * {@inheritDoc Vector.toArray}
   */
  public toArray(): S[] {
    return [...this._data];
  }

  /**
   * {@inheritDoc Vector.getSparseData}
   */
  public getSparseData(): Map<number, S> {
    const ops = this.ops();
    const zero = ops.zero();

    const sparseData: Map<number, S> = new Map();
    this.toArray().forEach((value, index) => {
      if (!ops.equals(zero, value)) {
        sparseData.set(index, value);
      }
    });

    return sparseData;
  }

  /**
   * {@inheritDoc Vector.getDimension}
   */
  public getDimension(): number {
    return this._data.length;
  }
}
