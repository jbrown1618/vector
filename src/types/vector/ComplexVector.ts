import { ArrayVector } from './ArrayVector';
import { ComplexNumber } from '../scalar/ComplexNumber';
import { VectorConstructor, VectorData } from './Vector';
import { ComplexMatrix, VectorBuilder } from '../../index';
import { StaticImplements } from '../../utilities/StaticImplements';
import { ComplexNumberOperations } from '../scalar/ComplexNumberOperations';

/**
 * Implements `Vector` as an array of `ComplexNumber`s
 */
@StaticImplements<VectorConstructor<ComplexNumber, ComplexVector>>()
export class ComplexVector extends ArrayVector<ComplexNumber> {
  static ops(): ComplexNumberOperations {
    return new ComplexNumberOperations();
  }

  static builder(): VectorBuilder<ComplexNumber, ComplexVector> {
    return new VectorBuilder(ComplexVector);
  }

  ops(): ComplexNumberOperations {
    return ComplexVector.ops();
  }

  builder(): VectorBuilder<ComplexNumber, ComplexVector> {
    return ComplexVector.builder();
  }

  matrixBuilder() {
    return ComplexMatrix.builder();
  }

  constructor(data: VectorData<ComplexNumber>) {
    super(data);
  }
}
