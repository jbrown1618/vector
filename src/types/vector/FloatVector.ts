import { StaticImplements } from '../../utilities/StaticImplements';
import { NumberOperations } from '../scalar/NumberOperations';
import { Vector, VectorConstructor, VectorData } from './Vector';
import { VectorBuilder } from './VectorBuilder';
import { MatrixBuilder } from '../matrix/MatrixBuilder';
import { assertValidVectorIndex, assertHomogeneous } from '../../utilities/ErrorAssertions';
import { Matrix } from '../matrix/Matrix';
import { FloatMatrix } from '../matrix/FloatMatrix';

/**
 * A dense {@link Vector} of `number`s implemented as a `Float64Array`
 * @public
 */
@StaticImplements<VectorConstructor<number, FloatVector>>()
export class FloatVector implements Vector<number> {
  public static ops(): NumberOperations {
    return new NumberOperations();
  }

  public static builder(): VectorBuilder<number, FloatVector> {
    return new VectorBuilder(FloatVector);
  }

  private _data: Float64Array;

  /**
   * @internal
   */
  constructor(data: VectorData<number> | Float64Array) {
    if (data instanceof Float64Array) {
      this._data = data;
    } else {
      this._data = Float64Array.from(data);
    }
  }

  /**
   * {@inheritDoc ArrayVector.ops}
   */
  public ops(): NumberOperations {
    return FloatVector.ops();
  }

  /**
   * {@inheritDoc ArrayVector.builder}
   */
  public builder(): VectorBuilder<number, FloatVector> {
    return FloatVector.builder();
  }

  /**
   * {@inheritDoc ArrayVector.matrixBuilder}
   */
  public matrixBuilder(): MatrixBuilder<number, FloatVector, FloatMatrix> {
    return FloatMatrix.builder();
  }

  /**
   * {@inheritDoc Vector.getEntry}
   */
  public getEntry(index: number): number {
    assertValidVectorIndex(this, index);
    return this._data[index];
  }

  /**
   * {@inheritDoc Vector.set}
   */
  public set(index: number, value: number): Vector<number> {
    assertValidVectorIndex(this, index);

    const newData = Float64Array.from(this._data);
    newData[index] = value;

    return new FloatVector(newData);
  }

  /**
   * {@inheritDoc Vector.add}
   */
  public add(other: Vector<number>): Vector<number> {
    assertHomogeneous([this, other]);

    const newData = new Float64Array(this._data);
    for (let i = 0; i < this.getDimension(); i++) {
      newData[i] = this._data[i] + other.getEntry(i);
    }

    return new FloatVector(newData);
  }

  /**
   * {@inheritDoc Vector.equals}
   */
  public equals(other: Vector<number>): boolean {
    if (this.getDimension() !== other.getDimension()) {
      return false;
    }

    const ops = this.ops();
    for (let i = 0; i < this.getDimension(); i++) {
      if (!ops.equals(this._data[i], other.getEntry(i))) return false;
    }

    return true;
  }

  /**
   * {@inheritDoc Vector.innerProduct}
   */
  public innerProduct(other: Vector<number>): number {
    assertHomogeneous([this, other]);

    let sum = 0;
    for (let i = 0; i < this.getDimension(); i++) {
      sum += this._data[i] * other.getEntry(i);
    }

    return sum;
  }

  /**
   * {@inheritDoc Vector.outerProduct}
   */
  public outerProduct(other: Vector<number>): Matrix<number> {
    const m = this.getDimension();
    const n = other.getDimension();

    if (!m || !n) return new FloatMatrix(new Float64Array(0), [0, 0]);

    const matrixData = new Float64Array(m * n);
    for (let j = 0; j < n; j++) {
      for (let i = 0; i < m; i++) {
        matrixData[j * m + i] = this._data[i] * other.getEntry(j);
      }
    }

    return new FloatMatrix(matrixData, [m, n]);
  }

  /**
   * {@inheritDoc Vector.projectOnto}
   */
  public projectOnto(u: Vector<number>): Vector<number> {
    const uDotU = u.innerProduct(u);
    if (uDotU === 0) throw new Error(`Cannot project onto the 0-vector`);
    const oneOverUDotU = 1 / uDotU;

    const uDotV = u.innerProduct(this);
    const magnitudeOfProjection = uDotV * oneOverUDotU;
    return u.scalarMultiply(magnitudeOfProjection);
  }

  /**
   * {@inheritDoc Vector.scalarMultiply}
   */
  public scalarMultiply(scalar: number): Vector<number> {
    return this.map(v => v * scalar);
  }

  /**
   * {@inheritDoc Vector.toArray}
   */
  public toArray(): number[] {
    return Array.from(this._data);
  }

  /**
   * {@inheritDoc Vector.getSparseData}
   */
  public getSparseData(): Map<number, number> {
    const sparseData: Map<number, number> = new Map();
    this.forEach((value, index) => {
      if (value !== 0) {
        sparseData.set(index, value);
      }
    });
    return sparseData;
  }

  /**
   * {@inheritDoc Vector.getDimension}
   */
  public getDimension(): number {
    return this._data.length;
  }

  /**
   * {@inheritDoc Vector.getDimension}
   */
  public forEach(callback: (entry: number, index: number) => void): void {
    this._data.forEach(callback);
  }

  /**
   * {@inheritDoc Vector.getDimension}
   */
  public map(valueFromEntry: (entry: number, index: number) => number): Vector<number> {
    return new FloatVector(this._data.map(valueFromEntry));
  }

  /**
   * {@inheritDoc Vector.getDimension}
   */
  public combine(
    other: Vector<number>,
    combineEntries: (a: number, b: number) => number
  ): Vector<number> {
    assertHomogeneous([this, other]);
    return this.map((entry, index) => combineEntries(entry, other.getEntry(index)));
  }
}
