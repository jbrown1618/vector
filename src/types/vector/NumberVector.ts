import { StaticImplements } from '../../utilities/StaticImplements';
import { NumberMatrix } from '../matrix/NumberMatrix';
import { NumberOperations } from '../scalar/NumberOperations';
import { ArrayVector } from './ArrayVector';
import { VectorConstructor, VectorData } from './Vector';
import { VectorBuilder } from './VectorBuilder';
import { MatrixBuilder } from '../matrix/MatrixBuilder';

/**
 * A `Vector` implemented as an array of JS `number` primitives.
 * @public
 */
@StaticImplements<VectorConstructor<number, NumberVector>>()
export class NumberVector extends ArrayVector<number> {
  public static ops(): NumberOperations {
    return new NumberOperations();
  }

  public static builder(): VectorBuilder<number, NumberVector> {
    return new VectorBuilder(NumberVector);
  }

  constructor(data: VectorData<number>) {
    super(data);
  }

  /**
   * {@inheritDoc TODO}
   */
  public ops(): NumberOperations {
    return NumberVector.ops();
  }

  /**
   * {@inheritDoc TODO}
   */
  public builder(): VectorBuilder<number, NumberVector> {
    return NumberVector.builder();
  }

  /**
   * {@inheritDoc TODO}
   */
  public matrixBuilder(): MatrixBuilder<number, NumberVector, NumberMatrix> {
    return NumberMatrix.builder();
  }
}
