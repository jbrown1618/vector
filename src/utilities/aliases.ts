import { Vector } from '@lib/types/vector/Vector';
import { Matrix } from '@lib/types/matrix/Matrix';
import { NumberVector } from '@lib/types/vector/NumberVector';
import { NumberMatrix } from '@lib/types/matrix/NumberMatrix';

const vb = NumberVector.builder();
const mb = NumberMatrix.builder();

/**
 * Creates a new {@link Vector} of numbers.  See {@link VectorBuilder.fromArray}
 * @public
 */
export function vec(data: number[]): Vector<number> {
  return vb.fromArray(data);
}

/**
 * Creates a new {@link Matrix} of numbers.  See {@link MatrixBuilder.fromArray}
 * @public
 */
export function mat(data: number[][]): Matrix<number> {
  return mb.fromArray(data);
}

/**
 * Creates a new identity matrix of size `size`.  See {@link MatrixBuilder.identity}
 * @public
 */
export function eye(size: number): Matrix<number> {
  return mb.identity(size);
}

/**
 * Creates a new vector of all 1s.  See {@link VectorBuilder.ones}
 * @public
 */
export function ones(entries: number): Vector<number>;
/**
 * Creates a new matrix of all 1s.  See {@link MatrixBuilder.ones}
 * @public
 */
export function ones(rows: number, columns: number): Matrix<number>;
export function ones(rows: number, columns?: number): Matrix<number> | Vector<number> {
  return columns === undefined ? vb.ones(rows) : mb.ones(rows, columns);
}

/**
 * Creates a new vector of all 0s.  See {@link VectorBuilder.zeros}
 * @public
 */
export function zeros(entries: number): Vector<number>;
/**
 * Creates a new matrix of all 0s.  See {@link MatrixBuilder.zeros}
 * @public
 */
export function zeros(rows: number, columns: number): Matrix<number>;
export function zeros(rows: number, columns?: number): Matrix<number> | Vector<number> {
  return columns === undefined ? vb.zeros(rows) : mb.zeros(rows, columns);
}

/**
 * Creates a new matrix with the specified entries on the diagonal.  See {@link MatrixBuilder.diagonal}
 * @public
 */
export function diag(elements: number[]): Matrix<number> {
  return mb.diagonal(vec(elements));
}
