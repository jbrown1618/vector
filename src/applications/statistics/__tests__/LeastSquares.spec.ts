import { vec, mat } from '@lib/utilities/aliases';
import { Vector } from '@lib/types/vector/Vector';
import {
  calculateGeneralLeastSquares,
  calculateLinearLeastSquares,
  solveOverdeterminedSystem
} from '../LeastSquares';
import { loadTestData } from '@test-utils/testData';

describe('LeastSquares', () => {
  const singleVariableTestData = loadTestData('least-squares-1d');
  const multiVariableTestData = loadTestData('least-squares-2d');

  describe('calculateLinearLeastSquaresApproximation', () => {
    test('calculates the coefficients for a simple linear regression', () => {
      const data = singleVariableTestData.getRowVectors();
      const result = calculateLinearLeastSquares(data);

      // According to Excel, the trend line equation for this data is y = -5.7147 + 3.2108x
      expect(result.coefficients).toMatchSnapshot();

      // The approximation function should be a line
      const expectedApproximator = (input: number) =>
        result.coefficients.getEntry(0) + result.coefficients.getEntry(1) * input;

      const inputValuesToCheck = [-1, 0, 1, 20];
      inputValuesToCheck.forEach(value => {
        const approximated = result.approximationFunction(vec([value]));
        const expected = expectedApproximator(value);
        expect(approximated).toEqual(expected);
      });
    });

    test('calculates the coefficients for a multiple linear regression', () => {
      const data = multiVariableTestData.getRowVectors();
      const result = calculateLinearLeastSquares(data);

      expect(result.coefficients).toMatchSnapshot();

      // The approximation function should be a plane
      const expectedApproximator = (x1: number, x2: number) =>
        result.coefficients.getEntry(0) +
        result.coefficients.getEntry(1) * x1 +
        result.coefficients.getEntry(2) * x2;

      const inputValuesToCheck = [-1, 0, 1, 20];
      inputValuesToCheck.forEach(value => {
        const approximated = result.approximationFunction(vec([value, value]));
        const expected = expectedApproximator(value, value);
        expect(approximated).toEqual(expected);
      });
    });

    test('yields an exact solution when there are two data points', () => {
      const exactData = [vec([0, 0]), vec([1, 1])];
      const result = calculateLinearLeastSquares(exactData);

      // For the data (0,0), (1,1), the exact solution is y=x
      const expectedCoefficients = vec([0, 1]);
      const expectedApproximator = (x: number) => x;

      expect(result.coefficients).toStrictEqual(expectedCoefficients);

      const inputValuesToCheck = [-1, 0, 1, 20];
      inputValuesToCheck.forEach(value => {
        const approximated = result.approximationFunction(vec([value]));
        const expected = expectedApproximator(value);
        expect(approximated).toEqual(expected);
      });
    });

    test('rejects non-homogeneous data', () => {
      const nonHomogeneousData = [vec([1, 1]), vec([1, 1, 1])];
      expect(() => calculateLinearLeastSquares(nonHomogeneousData)).toThrow();
    });

    test('rejects empty data', () => {
      expect(() => calculateLinearLeastSquares([])).toThrow();
    });
  });

  describe('calculateGeneralLeastSquaresApproximation', () => {
    test('calculates a quadratic regression', () => {
      const data = singleVariableTestData.getRowVectors();
      const quadraticTemplate = (coefficients: Vector<number>) => (inputs: Vector<number>) => {
        const x = inputs.getEntry(0);
        const constantTerm = coefficients.getEntry(0);
        const linearTerm = coefficients.getEntry(1) * x;
        const quadraticTerm = coefficients.getEntry(2) * Math.pow(x, 2);
        return constantTerm + linearTerm + quadraticTerm;
      };
      const result = calculateGeneralLeastSquares(data, quadraticTemplate, 3);

      // According to Excel, the trend line equation for this data is y = 3.42 + 0.35x + 0.16x^2
      expect(result.coefficients).toMatchSnapshot();

      // The approximation function should be a parabola
      const expectedApproximator = (input: number) =>
        result.coefficients.getEntry(0) +
        result.coefficients.getEntry(1) * input +
        result.coefficients.getEntry(2) * Math.pow(input, 2);

      const inputValuesToCheck = [-1, 0, 1, 20];
      inputValuesToCheck.forEach(value => {
        const approximated = result.approximationFunction(vec([value]));
        const expected = expectedApproximator(value);
        expect(approximated).toEqual(expected);
      });
    });

    const uselessFunctionTemplate = () => () => 0;

    test('rejects non-homogeneous data', () => {
      const nonHomogeneousData = [vec([1, 1]), vec([1, 1, 1])];
      expect(() =>
        calculateGeneralLeastSquares(nonHomogeneousData, uselessFunctionTemplate, 0)
      ).toThrow();
    });

    test('rejects empty data', () => {
      expect(() => calculateGeneralLeastSquares([], uselessFunctionTemplate, 0)).toThrow();
    });
  });

  describe('solveOverdeterminedSystem', () => {
    test('gives an approximate solution to an overdetermined system', () => {
      const A = loadTestData('random-50x3');
      const b = loadTestData('random-50x1').getColumn(0);
      const x = solveOverdeterminedSystem(A, b);
      expect(x).toMatchSnapshot();
    });

    test('rejects a system where A is underdetermined', () => {
      const A = mat([[1, 2, 3], [4, 5, 6]]);
      const b = vec([1, 2]);

      const solution = solveOverdeterminedSystem(A, b);
      expect(solution).not.toBeUndefined;
    });

    test('rejects a system with a dimension mismatch', () => {
      const A = mat([[1, 2], [3, 4], [5, 6]]);
      const b = vec([1, 2, 3, 4]);

      expect(() => solveOverdeterminedSystem(A, b)).toThrow();
    });
  });
});
