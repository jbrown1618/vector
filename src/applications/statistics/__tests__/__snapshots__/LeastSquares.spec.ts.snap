// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeastSquares calculateGeneralLeastSquaresApproximation calculates a quadratic regression 1`] = `
FloatVector {
  "_data": Float64Array [
    3.4235294117647626,
    0.3250257997935815,
    0.1603199174406616,
  ],
}
`;

exports[`LeastSquares calculateLinearLeastSquaresApproximation calculates the coefficients for a multiple linear regression 1`] = `
FloatVector {
  "_data": Float64Array [
    -0.40833333333332367,
    0.7366666666666807,
    2.5416666666666448,
  ],
}
`;

exports[`LeastSquares calculateLinearLeastSquaresApproximation calculates the coefficients for a simple linear regression 1`] = `
FloatVector {
  "_data": Float64Array [
    -5.7147058823529315,
    3.210784313725489,
  ],
}
`;

exports[`LeastSquares solveOverdeterminedSystem gives an approximate solution to an overdetermined system 1`] = `
FloatVector {
  "_data": Float64Array [
    0.07864000371772194,
    -0.10783427997814247,
    -0.0254725919314209,
  ],
}
`;
